### DISCUSSION TITLE
Rule not fired if ZbReceivedTopic is On

### DISCUSSION THREAD

**Discussion by webfrank** [arendst/Tasmota#15751] [Q&A] [answered]
### PROBLEM DESCRIPTION
_A clear and concise description of what the problem is._

If I set ZbReceivedTopic to ON all of my rules with on zbreceived#sensor does not trigger anymore. 

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [X ] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [ X] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [X ] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [X ] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [X ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [X ] Device used (e.g., Sonoff Basic): Sonoff ZbBridge
- [ ] Tasmota binary firmware version number used: 11.1.0
  - [X ] Pre-compiled
  - [ ] Self-compiled
- [ X] Flashing tools used: esptool
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:

Empty
```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:

stat/zbbridge/RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":171,"Free":340,"Rules":"on zbreceived#control#power=0 do zbsend {\"device\":\"relay1\", \"send\":{\"power\":0}} endon on zbreceived#control#power=1 do zbsend {\"device\":\"relay1\", \"send\":{\"power\":1}} endon"}}
18:11:32.564 MQT: stat/zbbridge/RESULT = {"Rule2":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":0,"Free":511,"Rules":""}}
18:11:32.809 MQT: stat/zbbridge/RESULT = {"Rule3":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":0,"Free":511,"Rules":""}}
```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
18:12:33.370 MQT: stat/zbbridge/STATUS = {"Status":{"Module":75,"DeviceName":"Tasmota","FriendlyName":["Tasmota"],"Topic":"zbbridge","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0}}
18:12:33.608 MQT: stat/zbbridge/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"Power On","Uptime":"0T04:08:22","StartupUTC":"2022-06-04T13:04:11","Sleep":50,"CfgHolder":4617,"BootCount":8,"BCResetTime":"2022-06-02T08:26:11","SaveCount":29,"SaveAddress":"FC000"}}
18:12:33.829 MQT: stat/zbbridge/STATUS2 = {"StatusFWR":{"Version":"11.1.0(zbbridge)","BuildDateTime":"2022-04-13T06:40:30","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":160,"Hardware":"ESP8266EX","CR":"395/699"}}
18:12:34.060 MQT: stat/zbbridge/STATUS3 = {"StatusLOG":{"SerialLog":0,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["CASA",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C80001000600003C5A00190000000000","00000080","40206082","00004000"]}}
18:12:34.306 MQT: stat/zbbridge/STATUS4 = {"StatusMEM":{"ProgramSize":678,"Free":1112,"Heap":18,"ProgramFlashSize":2048,"FlashSize":2048,"FlashChipId":"1540A1","FlashFrequency":40,"FlashMode":3,"Features":["00000809","0F1007C6","04400001","00000003","00000000","00000000","00020080","00200000","04000000"],"Drivers":"1,2,4,7,9,10,12,20,23,38,41,50,56,62","Sensors":"1"}}
18:12:34.545 MQT: stat/zbbridge/STATUS5 = {"StatusNET":{"Hostname":"zbbridge-0995","IPAddress":"192.168.100.129","Gateway":"192.168.100.1","Subnetmask":"255.255.255.0","DNSServer1":"192.168.100.1","DNSServer2":"0.0.0.0","Mac":"C4:5B:BE:E2:23:E3","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
18:12:34.778 MQT: stat/zbbridge/STATUS6 = {"StatusMQT":{"MqttHost":"e15c623f0bb1498d959e44b32d038435.s1.eu.hivemq.cloud","MqttPort":8883,"MqttClientMask":"zbbridge","MqttClient":"zbbridge","MqttUser":"tasmota","MqttCount":2,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
18:12:35.014 MQT: stat/zbbridge/STATUS7 = {"StatusTIM":{"UTC":"2022-06-04T17:12:34","Local":"2022-06-04T18:12:34","StartDST":"2022-03-27T02:00:00","EndDST":"2022-10-30T03:00:00","Timezone":"+01:00","Sunrise":"04:49","Sunset":"20:47"}}
18:12:35.256 MQT: stat/zbbridge/STATUS10 = {"StatusSNS":{"Time":"2022-06-04T18:12:35"}}
18:12:35.485 MQT: stat/zbbridge/STATUS11 = {"StatusSTS":{"Time":"2022-06-04T18:12:35","Uptime":"0T04:08:24","UptimeSec":14904,"Vcc":3.505,"Heap":18,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":2,"Wifi":{"AP":1,"SSId":"CASA","BSSId":"46:D5:F2:2C:DD:CF","Channel":1,"Mode":"11n","RSSI":34,"Signal":-83,"LinkCount":1,"Downtime":"0T00:00:05"}}}

```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Steps to reproduce the behavior:_

Enable ZbReceivedTopic 1

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

Rule on zbreceived should trigger

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by barbudor** [comment, marked as answer]
When enabling ZbReceivedTopic, the MQTT message change structure
So you need to adapt your rules

With OFF:
```
{"ZbReceived":{"Device":"0x35D7","Name":"Door1","0500<00":"010000010000","ZoneStatusChange":1,"ZoneStatusChangeZone":1,"Contact":1,"Endpoint":1,"LinkQuality":162}}
```
Rule trigger is `ON ZbReceived#Device DO `

With ON
```
{"Device":"0x35D7","Name":"Door1","0500<00":"000000010000","ZoneStatusChange":0,"ZoneStatusChangeZone":1,"Contact":0,"Endpoint":1,"LinkQuality":167}
```
Rule trigger is `ON Device... DO`

So you have to adapt your trigger, removing the `ZbReceived` part
This is for my door sensor.
Adapt to your case

**Reply by webfrank** [reply]
Hi, thank you for the clarification. Probably a note in the documentation would help.

**Reply by sfromis** [reply]
The general idea is that people know how the JSON-based triggers in rules work. Anyway, if you primarily want to process incoming data using rules, using `ZbReceivedTopic` makes it troublesome to create rules depending on which device it is. What I'd suggest, as a starting point is:
1. Always use ZbName to not depend on those "less readable" short addresses.
2. Use ZbNameKey to have that friendly name used as JSON key.
3. Use ZbNoPrefix to avoid that extra `ZbReceived` level in the JSON.

This allows nicely straightforward rules like
`on Ikea-Opal#Dimmer<10 do .... endon`
